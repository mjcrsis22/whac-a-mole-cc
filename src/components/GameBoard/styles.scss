.game-board-container {
  background-image: url(./backgrounds/chevrons-ground.svg);
  border-radius: 0.5rem;
  padding: 1rem;

  .play-container {
    position: relative;
    width: 100%;
    height: 15rem;

    .spot-wrapper {
      position: absolute;
    }
  }
}
