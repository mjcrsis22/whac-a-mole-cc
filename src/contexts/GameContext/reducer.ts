import {
  GAME_LEVEL,
  GameSetting,
  GAME_SETTINGS_BY_LEVEL,
} from '../../constants/game';

export interface State {
  started: boolean;
  levelSetting: GameSetting;
  points: number;
}

export const INITIAL_STATE: State = {
  started: false,
  levelSetting: GAME_SETTINGS_BY_LEVEL[GAME_LEVEL.LOW],
  points: 0,
};

export enum ActionTypes {
  SET_STARTED = 'SET_STARTED',
  SET_LEVEL = 'SET_LEVEL',
  SET_POINTS = 'SET_POINTS',
  SET_DEFAULT = 'SET_DEFAULT',
}

export type SetStarted = {
  type: ActionTypes.SET_STARTED;
  payload: boolean;
};

export type SetLevel = {
  type: ActionTypes.SET_LEVEL;
  payload: GameSetting;
};

export type SetPoints = {
  type: ActionTypes.SET_POINTS;
  payload: number;
};

export type SetDefault = {
  type: ActionTypes.SET_DEFAULT;
};

export type Actions = SetStarted | SetLevel | SetPoints | SetDefault;

export function reducer(state: State, action: Actions) {
  switch (action.type) {
    case ActionTypes.SET_STARTED:
      return {
        ...state,
        started: action.payload,
      };
    case ActionTypes.SET_LEVEL:
      return {
        ...state,
        levelSetting: action.payload,
      };
    case ActionTypes.SET_POINTS:
      return {
        ...state,
        points: action.payload,
      };
    default:
      return state;
  }
}
