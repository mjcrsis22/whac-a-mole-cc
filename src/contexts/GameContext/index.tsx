import { createContext, Dispatch, ReactNode, useReducer } from 'react';
import { State, INITIAL_STATE, Actions, reducer } from './reducer';

export interface GameContextType {
  state: State;
  dispatch: Dispatch<Actions>;
}

// eslint-disable-next-line react-refresh/only-export-components
export const GameContext = createContext<GameContextType>({
  state: INITIAL_STATE,
  dispatch: () => {},
});

interface GameContextProviderProps {
  children: ReactNode;
}

export default function GameContextProvider({
  children,
}: GameContextProviderProps) {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);

  return (
    <GameContext.Provider value={{ state, dispatch }}>
      {children}
    </GameContext.Provider>
  );
}
