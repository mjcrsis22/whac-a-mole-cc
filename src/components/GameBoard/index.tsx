import { useCallback, useContext, useEffect, useState } from 'react';
import { GameContext } from '../../contexts/GameContext';
import { ActionTypes } from '../../contexts/GameContext/reducer';
import { SPOT_POSITIONS, SPOT_STATES } from '../../constants/spot';
import GameSpot from '../GameSpot';
import './styles.scss';

interface BoardSpot {
  left: string;
  top: string;
  state: SPOT_STATES;
}

function GameBoard() {
  const {
    state: { started, levelSetting },
    dispatch,
  } = useContext(GameContext);
  const [boardSpots, setBoardSpots] = useState<BoardSpot[]>([]);
  const [threadId, setThreadId] = useState<number>(0);

  const setActiveSpot = (spotIdx: number) => {
    setBoardSpots(state =>
      state.map((spot, idx) => ({
        ...spot,
        state:
          idx === spotIdx && spot.state === SPOT_STATES.inactive
            ? SPOT_STATES.active
            : SPOT_STATES.inactive,
      })),
    );
  };

  const activateRandomSpot = useCallback(() => {
    setActiveSpot(Math.floor(Math.random() * boardSpots.length));
  }, [boardSpots.length]);

  const restartGame = useCallback(() => {
    activateRandomSpot();
    setThreadId(setTimeout(restartGame, levelSetting.spotRate));
  }, [activateRandomSpot, levelSetting.spotRate]);

  const stopGame = useCallback(() => {
    clearTimeout(threadId);
    setActiveSpot(-1);
    setThreadId(0);
  }, [threadId]);

  useEffect(() => {
    setBoardSpots(
      SPOT_POSITIONS.map(position => ({
        ...position,
        state: SPOT_STATES.inactive,
      })),
    );
  }, []);

  useEffect(() => {
    if (started && threadId === 0) {
      dispatch({ type: ActionTypes.SET_POINTS, payload: 0 });
      restartGame();
    } else if (!started && threadId !== 0) {
      stopGame();
    }
  }, [restartGame, stopGame, started, threadId, dispatch]);

  useEffect(() => {
    clearTimeout(threadId);
    restartGame();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [levelSetting]);

  return (
    <div className="game-board-container">
      <div className="play-container">
        {boardSpots.map(({ left, top, state }) => (
          <div
            key={`${left}-${top}`}
            className="spot-wrapper"
            style={{ left, top }}
          >
            <GameSpot state={state} />
          </div>
        ))}
      </div>
    </div>
  );
}

export default GameBoard;
