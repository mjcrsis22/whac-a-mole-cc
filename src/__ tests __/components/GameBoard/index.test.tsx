import { useContext } from 'react';
import { render } from '@testing-library/react';
import GameContextProvider from '../../../contexts/GameContext';
import { INITIAL_STATE } from '../../../contexts/GameContext/reducer';
import GameBoard from '../../../components/GameBoard';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: jest.fn(),
}));

const mockedContext = {
  state: INITIAL_STATE,
  dispatch: jest.fn(),
};

const mockedUseContext = useContext as jest.MockedFn<typeof useContext>;

describe('GameBoard component', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  beforeEach(() => {
    mockedUseContext.mockReturnValue(mockedContext);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test.each([true, false])(
    'Renders correctly - according to the started value',
    async started => {
      mockedUseContext.mockReturnValue({
        ...mockedContext,
        state: { ...mockedContext.state, started },
      });
      const { container } = render(
        <GameContextProvider>
          <GameBoard />
        </GameContextProvider>,
      );
      expect(container).toMatchSnapshot();
    },
  );
});
