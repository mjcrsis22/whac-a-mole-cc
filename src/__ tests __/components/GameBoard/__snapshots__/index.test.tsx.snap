// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBoard component Renders correctly - according to the started value 1`] = `
<div>
  <div
    class="game-board-container"
  >
    <div
      class="play-container"
    >
      <div
        class="spot-wrapper"
        style="left: 5%; top: 0%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 5%; top: 72%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 30%; top: 15%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 30%; top: 60%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 45%; top: 40%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 60%; top: 15%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 60%; top: 60%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 85%; top: 0%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 85%; top: 72%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`GameBoard component Renders correctly - according to the started value 2`] = `
<div>
  <div
    class="game-board-container"
  >
    <div
      class="play-container"
    >
      <div
        class="spot-wrapper"
        style="left: 5%; top: 0%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 5%; top: 72%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 30%; top: 15%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 30%; top: 60%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 45%; top: 40%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 60%; top: 15%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 60%; top: 60%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 85%; top: 0%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="spot-wrapper"
        style="left: 85%; top: 72%;"
      >
        <div
          class="game-spot"
        >
          <div
            class="spot-mark "
          />
          <div
            class="spot-mole-wrapper"
          >
            <img
              alt="Mole"
              class="spot-mole "
              role="button"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
